pkg/sinks/gcppubsub/gcppubsub.go

```diff
@@ -0,0 +1,96 @@
+package gcppubsub
+
+import (
+	"context"
+	"flag"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	go_metrics "github.com/kentik/go-metrics"
+	"github.com/kentik/ktranslate/pkg/eggs/logger"
+	"github.com/kentik/ktranslate/pkg/formats"
+	"github.com/kentik/ktranslate/pkg/kt"
+)
+
+type GCPPubSub struct {
+	logger.ContextL
+	client    *pubsub.Client
+	registry  go_metrics.Registry
+	metrics   *GCPPubSubMetric
+	projectID string
+	topicID   string
+	topic     *pubsub.Topic
+}
+
+type GCPPubSubMetric struct {
+	DeliveryErr go_metrics.Meter
+	DeliveryWin go_metrics.Meter
+}
+
+var (
+	GCPPubSubProjectID = flag.String("gcp_pubsub_project_id", "", "GCP PubSub Project ID to use")
+	GCPPubSubTopic     = flag.String("gcp_pubsub_topic", "", "GCP PubSub Topic to publish to")
+)
+
+func NewSink(log logger.Underlying, registry go_metrics.Registry) (*GCPPubSub, error) {
+	return &GCPPubSub{
+		registry:  registry,
+		projectID: *GCPPubSubProjectID,
+		topicID:   *GCPPubSubTopic,
+		ContextL:  logger.NewContextLFromUnderlying(logger.SContext{S: "gcppubsub"}, log),
+		metrics: &GCPPubSubMetric{
+			DeliveryErr: go_metrics.GetOrRegisterMeter("delivery_errors_gcppubsub", registry),
+			DeliveryWin: go_metrics.GetOrRegisterMeter("delivery_wins_gcppubsub", registry),
+		},
+	}, nil
+}
+
+func (s *GCPPubSub) Init(ctx context.Context, format formats.Format, compression kt.Compression, fmtr formats.Formatter) error {
+	if s.projectID == "" {
+		return fmt.Errorf("No project ID was set for GCP PubSub. Use the -gcp_pubsub_project_id option to set one up.")
+	}
+	if s.topicID == "" {
+		return fmt.Errorf("No topic was set for GCP PubSub. Use the -gcp_pubsub_topic option to set one up.")
+	}
+	client, err := pubsub.NewClient(ctx, s.projectID)
+	if err != nil {
+		return fmt.Errorf("There was an error when creating the gcp pubsub client: %v.", err)
+	}
+	s.client = client
+	s.topic = s.client.Topic(s.topicID)
+
+	s.Infof("connected to GCP PubSub for project %s on topic %s", s.projectID, s.topicID)
+
+	return nil
+}
+
+func (s *GCPPubSub) Send(ctx context.Context, payload *kt.Output) {
+	go s.send(ctx, payload.Body)
```

Comment:
I use this pattern everywhere but I worry that in an extreme case there will be too many goroutines running. I don't think worth worrying about it now.

Reply:
I thought about this too but I think we can revisit later.  Maybe we can add some profiling as well to get some more data.